File: client2/src/types/expense.ts
export type RecurringType = 'none' | 'fixed_day' | 'flexible'
export type RecurringPeriod =
  | 'none'
  | 'weekly'
  | 'biweekly'
  | 'monthly'
  | 'bimonthly'
  | 'quarterly'
  | 'fourmonths'
  | 'semiannually'
  | 'annually'

export interface ExpenseTypeParentSummary {
  id: number
  name: string
  icon: string
  color: string
}

export interface ExpenseTypeWalletSummary {
  id: number
  name: string
  icon: string
  color: string
  is_credit: boolean
  is_cash: boolean
}

export interface ExpenseType {
  id: number
  parent_id?: number | null
  name: string
  icon: string
  color: string
  description: string
  default_amount: number
  default_wallet_id?: number | null
  recurring_type: RecurringType
  recurring_period: RecurringPeriod
  recurring_due_day: number
  next_due_day?: string | null
  stopped: boolean
  user_id: number
  created_at: string
  updated_at: string
  parent?: ExpenseTypeParentSummary | null
  default_wallet?: ExpenseTypeWalletSummary | null
}

export interface CreateExpenseTypeRequest {
  parent_id?: number | null
  name: string
  icon?: string
  color?: string
  description?: string
  default_amount?: number
  default_wallet_id?: number | null
  recurring_type?: RecurringType
  recurring_period?: RecurringPeriod
  recurring_due_day?: number
  next_due_day?: string | null
  stopped?: boolean
}

export type UpdateExpenseTypeRequest = CreateExpenseTypeRequest

export interface PostponeExpenseTypeRequest {
  next_due_day: string
}

export interface ExpenseTypeListResponse {
  expense_types: ExpenseType[]
  total: number
}

export interface ExpenseTypeTreeNode {
  expense_type: ExpenseType
  children: ExpenseType[]
}

export interface ExpenseTypeTreeResponse {
  tree: ExpenseTypeTreeNode[]
  total: number
}

export interface ExpenseWalletSummary {
  id: number
  name: string
  icon: string
  color: string
  is_credit: boolean
  is_cash: boolean
}

export interface ExpensePaymentSummary {
  id: number
  wallet_id: number
  amount: number
  date: string
  note: string
}

export interface Expense {
  id: number
  expense_type_id: number
  wallet_id?: number | null
  payment_id?: number | null
  amount: number
  date: string
  note: string
  user_id: number
  created_at: string
  updated_at: string
  expense_type?: ExpenseType | null
  wallet?: ExpenseWalletSummary | null
  payment?: ExpensePaymentSummary | null
}

export interface CreateExpenseRequest {
  expense_type_id: number
  wallet_id?: number | null
  payment_id?: number | null
  amount: number
  date: string
  note?: string
}

export type UpdateExpenseRequest = CreateExpenseRequest

export interface ExpenseListResponse {
  expenses: Expense[]
  total: number
}

File: server/internal/expenses/expense.go
package expenses

import (
	"time"

	"gorm.io/gorm"
)

type Expense struct {
	ID            uint           `json:"id" gorm:"primaryKey;type:bigint"`
	ExpenseTypeID uint           `json:"expense_type_id" gorm:"type:bigint;not null;index"`
	WalletID      *uint          `json:"wallet_id" gorm:"type:bigint;index"`
	PaymentID     *uint          `json:"payment_id" gorm:"type:bigint;index"`
	Amount        float64        `json:"amount" gorm:"type:numeric(12,2);not null"`
	Date          time.Time      `json:"date" gorm:"type:date;not null;index"`
	Note          string         `json:"note" gorm:"type:text"`
	UserID        uint           `json:"user_id" gorm:"type:bigint;not null;index"`
	CreatedAt     time.Time      `json:"created_at"`
	UpdatedAt     time.Time      `json:"updated_at"`
	DeletedAt     gorm.DeletedAt `json:"-" gorm:"index"`

	ExpenseType ExpenseType `json:"expense_type,omitempty" gorm:"foreignKey:ExpenseTypeID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
	Wallet      Wallet      `json:"wallet,omitempty" gorm:"foreignKey:WalletID;constraint:OnUpdate:CASCADE,OnDelete:SET NULL;"`
	Payment     Payment     `json:"payment,omitempty" gorm:"foreignKey:PaymentID;constraint:OnUpdate:CASCADE,OnDelete:SET NULL;"`
}
